File: Microsoft Office Developer Documentation Team/SharePoint 2010 101 Code Samples/49862-SharePoint 2010 101 Code Samples/SharePoint 2010 Creating Custom SharePoint Service Applications/VB/WCF_CustomServiceApplication/Features/Feature1/Feature1.EventReceiver.vb
Option Explicit On
Option Strict On

Imports System
Imports System.Runtime.InteropServices
Imports System.Security.Permissions
Imports Microsoft.SharePoint
Imports Microsoft.SharePoint.Security
Imports Microsoft.SharePoint.Administration
Imports WCF_CustomServiceApplication.Server
Imports WCF_CustomServiceApplication.Client

Namespace Features

    <GuidAttribute("eedc5939-6c90-4375-9ee1-4a2c7630c0cb")> _
    Public Class Feature1EventReceiver
        Inherits SPFeatureReceiver

        Public Overrides Sub FeatureActivated(ByVal properties As SPFeatureReceiverProperties)
            'install the service
            Dim service As DayNamerService = SPFarm.Local.Services.GetValue(Of DayNamerService)()
            If service Is Nothing Then
                service = New DayNamerService(SPFarm.Local)
                service.Update()
            End If

            'install the service proxy
            Dim serviceProxy As DayNamerServiceProxy = SPFarm.Local.ServiceProxies.GetValue(Of DayNamerServiceProxy)()
            If serviceProxy Is Nothing Then
                serviceProxy = New DayNamerServiceProxy(SPFarm.Local)
                serviceProxy.Update(True)
            End If

            'with service added to the farm, install instance
            Dim serviceInstance As DayNamerServiceInstance = New DayNamerServiceInstance(SPServer.Local, service)
            serviceInstance.Update(True)
        End Sub

        Public Overrides Sub FeatureDeactivating(ByVal properties As SPFeatureReceiverProperties)
            'uninstall the instance
            Dim serviceInstance As DayNamerServiceInstance = SPFarm.Local.Services.GetValue(Of DayNamerServiceInstance)()
            If serviceInstance IsNot Nothing Then
                SPServer.Local.ServiceInstances.Remove(serviceInstance.Id)
            End If
            'uninstall the service proxy
            Dim serviceProxy As DayNamerServiceProxy = SPFarm.Local.ServiceProxies.GetValue(Of DayNamerServiceProxy)()
            If serviceProxy IsNot Nothing Then
                SPFarm.Local.ServiceProxies.Remove(serviceProxy.Id)
            End If
            'uninstall the service
            Dim service As DayNamerService = SPFarm.Local.Services.GetValue(Of DayNamerService)()
            If service IsNot Nothing Then
                SPFarm.Local.Services.Remove(service.Id)
            End If
        End Sub

    End Class

End Namespace

File: Microsoft Office Developer Documentation Team/SharePoint 2010 101 Code Samples/[C#]-SharePoint 2010 101 Code Samples/SharePoint 2010 Creating Custom SharePoint Service Applications/C#/WCF_CustomServiceApplication/Features/Feature1/Feature1.EventReceiver.cs
using System;
using System.Runtime.InteropServices;
using System.Security.Permissions;
using Microsoft.SharePoint;
using Microsoft.SharePoint.Security;
using Microsoft.SharePoint.Administration;
using WCF_CustomServiceApplication.Server;
using WCF_CustomServiceApplication.Client;

namespace WCF_CustomServiceApplication.Features.Feature1
{

    [Guid("b8663c97-18cf-44ef-ab9a-42f04d64d7fd")]
    public class Feature1EventReceiver : SPFeatureReceiver
    {
        public override void FeatureActivated(SPFeatureReceiverProperties properties)
        {
            // install the service
            DayNamerService service = SPFarm.Local.Services.GetValue<DayNamerService>();
            if (service == null)
            {
                service = new DayNamerService(SPFarm.Local);
                service.Update();
            }

            // install the service proxy
            DayNamerServiceProxy serviceProxy = SPFarm.Local.ServiceProxies.GetValue<DayNamerServiceProxy>();
            if (serviceProxy == null)
            {
                serviceProxy = new DayNamerServiceProxy(SPFarm.Local);
                serviceProxy.Update(true);
            }

            // with service added to the farm, install instance
            DayNamerServiceInstance serviceInstance = new DayNamerServiceInstance(SPServer.Local, service);
            serviceInstance.Update(true);

        }


        public override void FeatureDeactivating(SPFeatureReceiverProperties properties)
        {
            // uninstall the instance
            DayNamerServiceInstance serviceInstance = SPFarm.Local.Services.GetValue<DayNamerServiceInstance>();
            if (serviceInstance != null)
                SPServer.Local.ServiceInstances.Remove(serviceInstance.Id);

            // uninstall the service proxy
            DayNamerServiceProxy serviceProxy = SPFarm.Local.ServiceProxies.GetValue<DayNamerServiceProxy>();
            if (serviceProxy != null)
            {
                SPFarm.Local.ServiceProxies.Remove(serviceProxy.Id);
            }

            // uninstall the service
            DayNamerService service = SPFarm.Local.Services.GetValue<DayNamerService>();
            if (service != null)
                SPFarm.Local.Services.Remove(service.Id);

        }
    }
}
